### PR TITLE
tests: implement tests for CRUD controller

### PR DESCRIPTION
**Pull request by gboutry**
First PR to implement tests into tdp_server

### CODE REVIEW

**Comment by rpignolet** [comment]
Need time to review it

File: tdp_server/conftest.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        "name": "mock_start",
+        "noop": True,
+        "depends_on": ["mock_config", "mock_node_start"],
+    },
+    {"name": "mock_init", "noop": True, "depends_on": ["mock_start", "mock_node_init"]},
+]
+
+
+@pytest.fixture
+def app(mock_runtime):
```
Comment (rpignolet) [inline tdp_server/conftest.py:41]: Can you put `mock_runtime` function/fixture before it is used ? I think it is more readable to read function before it is used.

File: tdp_server/conftest.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    app = create_app()
+    # Disable token validation for tests
+
+    dag, service_managers = mock_runtime
+    app.dependency_overrides[validate_token] = mock_validate_token
```
Comment (rpignolet) [inline tdp_server/conftest.py:46]: The comment should be just before the code. The line with `dag, service_managers = mock_runtime` is not used to disable token validation.
```suggestion

    dag, service_managers = mock_runtime
    # Disable token validation for tests
    app.dependency_overrides[validate_token] = mock_validate_token
```

File: tdp_server/api/v1/endpoints/test_component.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def test_get_component(client):
+    resp = client.get(f"{settings.API_V1_STR}/service/mock/component/node")
+    assert resp.status_code == status.HTTP_200_OK
+    response = resp.json()
+    assert response["id"] == "node"
+
+
+def test_get_component_does_exist(client):
```
Comment (rpignolet) [inline tdp_server/api/v1/endpoints/test_component.py:13]: The test assert that status code is 404 so:
```suggestion
def test_get_component_does_not_exist(client):
```

File: tdp_server/api/v1/endpoints/test_component.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    resp = client.put(f"{settings.API_V1_STR}/service/mock/component/node", json=data)
+    assert resp.status_code == status.HTTP_200_OK
+    response = resp.json()
+    assert data["message"] in response["message"]
+
+    resp = client.get(f"{settings.API_V1_STR}/service/mock/component/node")
+    response = resp.json()
+    assert response["variables"] == data["variables"]
+    print(response)
```
Comment (rpignolet) [inline tdp_server/api/v1/endpoints/test_component.py:43]: Print to remove.

**Review by rpignolet — CHANGES_REQUESTED** [review]

File: tdp_server/api/v1/endpoints/test_component.py (hunk 3 above)
Comment (gboutry) [reply tdp_server/api/v1/endpoints/test_component.py]: Done

File: tdp_server/api/v1/endpoints/test_component.py (hunk 4 above)
Comment (gboutry) [reply tdp_server/api/v1/endpoints/test_component.py]: Done

File: tdp_server/api/v1/endpoints/test_service.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def test_get_service(client):
+    resp = client.get(f"{settings.API_V1_STR}/service/mock")
+    assert resp.status_code == status.HTTP_200_OK
+    response = resp.json()
+    assert response["id"] == "mock"
+
+
+def test_get_service_does_exist(client):
```
[the remark this answers is not in the document]
Comment (gboutry) [reply tdp_server/api/v1/endpoints/test_service.py]: Done

File: tdp_server/api/dependencies.py [hunk 6]
```diff
@@ -33,31 +33,29 @@
...
+@lru_cache()
 def get_cluster_variables() -> ClusterVariables:
-    return CLUSTER_VARIABLES
+    return ClusterVariables.get_cluster_variables(settings.TDP_VARS)
 
 
-def get_runner_service() -> RunnerService:
-    return RUNNER_SERVICE
+@lru_cache()
+def get_runner_service(dag: Dag = Depends(get_dag)) -> RunnerService:
```
Comment (rpignolet) [inline tdp_server/api/dependencies.py:53]: To use `@lru_cache()`, argument must be hashable, a `Dag` is hashable ?

File: tdp_server/api/dependencies.py [hunk 7]
```diff
@@ -33,31 +33,29 @@
...
-)
-CLUSTER_VARIABLES = ClusterVariables.get_cluster_variables(settings.TDP_VARS)
-
 
 def get_db() -> Generator:
     with SessionLocal() as db:
         yield db
 
 
+@lru_cache()
```
Comment (rpignolet) [inline tdp_server/api/dependencies.py:42]: This function have no argument, should we use `cache()` instead ?

Comment (rpignolet) [reply tdp_server/api/dependencies.py:42]: Never mind, `cache()` added in python 3.9. We must support python 3.6.

File: tdp_server/api/dependencies.py (hunk 6 above)
Comment (gboutry) [reply tdp_server/api/dependencies.py:53]: It's the function arguments that must be hashable, not the result

Comment (rpignolet) [reply tdp_server/api/dependencies.py:53]: Yes and the function argument is an instance of `Dag` no ?

Comment (rpignolet) [reply tdp_server/api/dependencies.py:53]: I think what you want is a singleton not a cache or maybe I don't understand the code.

Comment (gboutry) [reply tdp_server/api/dependencies.py:53]: There's no implementation of singleton in python. `lru_cache` for this kind of behavior is the accepted practice.

My goal was to defer `Dag`, `ClusterVariables`, and `RunnerService` instanciation to be able to replace them easily in the tests.

**Review by rpignolet — APPROVED** [review]

File: tdp_server/api/dependencies.py (hunk 6 above)
Comment (gboutry) [reply tdp_server/api/dependencies.py:53]: Yeah, from what I research on the internet, the default `hash` value is `id(X)//16` (id being the memory address)

**Pull request merged**